### DISCUSSION TITLE
How to use GitHub Actions to pull specfic files from someone else's repository when they get updates?

### DISCUSSION THREAD

**Discussion by lwcorp** [community/community#53762] [Actions] [answered]
### Select Topic Area

Question

### Body

I'd like to utilize Github Actions to pull specific files from another repo whenever they get updated.

What I tried to do is:

1. Pull specific files from someone else's repo (and use a token if it's a private repo).
1. Rename some of those files.
1. Create a PR with them in my repo.
1. Allow to control whether to do manually or also automatically (when the source repo updates itself).

Keep in mind it needs to use the latest Node version due to Github's blockage of older Node versions.

Sample codes (trying to pull from a public repo) are below (with v3 in actions/checkout@v3 in an attempt to use the latest Node), but how to make it trigger when someone else's repo updates its files?

Sample code:

> git add $FILES_NEW
>   git commit -m "Pulled files from $SOURCE_REPO."
>   shell: /usr/bin/bash -e {0}
> Author identity unknown
> 
> *** Please tell me who you are.
> 
> Run
> 
>   git config --global user.email "you@example.com"
>   git config --global user.name "Your Name"
> 
> to set your account's default identity.
> Omit --global to set the identity only in this repository.
> 
> fatal: empty ident name (for <runner@...>) not allowed
> Error: Process completed with exit code 128.
> 

```
name: Pull File from Another Repository on Push
env:
  SOURCE_REPO: https://github.com/some/repo
  AUTOMATIC_MONITOR: true
  FILES: "*.txt,LICENSE" # seperate multiple entries with ,
  FILES_NEW: "*.txt,LICENSE*" # seperate multiple entries with ,
  EXTRA: mv LICENSE LICENSE_some_repo # do an extra command like renaming
  # TOKEN_NAME: ACCESS_TOKEN # uncomment to use a secret with that name
  # SOURCE_BRANCH: source_alternate # uncomment to not monitor the default branch
  # TARGET_BRANCH: target_alternate # uncomment to not push to default branch

on: # Remove # below for non default branches
  workflow_dispatch:
#    inputs:
#      source_branch:
#        description: '${{ env.SOURCE_BRANCH }}'
#        required: false
  push:
#    branches:
#      - ${{ env.SOURCE_BRANCH }}
    paths:
      - ./source-repo/**

jobs:
  pull-file:
    runs-on: ubuntu-latest
    steps:
      - name: Check whether to automatically monitor
        if: ${{ github.event_name != 'workflow_dispatch' && env.AUTOMATIC_MONITOR == false }}
        run: |
          exit 1
      - name: Checkout
        uses: actions/checkout@v3

      - name: Pull File
        env:
          # THE_TOKEN: ${{ secrets.$TOKEN_NAME }}
          EVENT_NAME: ${{ github.event_name }}
          SOURCE_BRANCH: ${{ github.event.inputs.source_branch }}
          REF: ${{ github.ref }}
        run: |
          #!/bin/sh
          
          if [ -n "$TOKEN_NAME" ]; then
            REPO_URL=https://$THE_TOKEN@$SOURCE_REPO.git
          else
            REPO_URL=$SOURCE_REPO.git
          fi
          
          if [ "$EVENT_NAME" = "workflow_dispatch" ]; then
            if [ -n "$SOURCE_BRANCH" ]; then
              git clone -b $SOURCE_BRANCH $REPO_URL source-repo
            else
              git clone $REPO_URL source-repo
            fi
          else
            git clone -b $REF $REPO_URL source-repo
          fi
          cd source-repo
          $EXTRA
          cp -u $FILES_NEW ../
          cd ..
      - name: Commit
        run: |
          git config user.name "$USERNAME"
          git config user.email "$USERNAME@$ADDRESS_SUFFIX"
          git add $FILES_NEW
          git commit -m "Pulled files from $SOURCE_REPO."
      - name: Push
        run: |
          if [ -n "$TARGET_BRANCH" ]; then
            git push origin $TARGET_BRANCH
          else
            git push
          fi
```

**Comment by airtower-luna** [comment]
In the first version, the problem is simply [`actions/github-script`](https://github.com/actions/github-script#actionsgithub-script) is for writing a step in *Javascript*. The script you give it is a shell script, which naturally isn't valid Javascript, so you get a syntax error.

Putting your shell script into a `run` step is the right approach. The problem with your `git clone` command is that you try to use `https://TOKEN@github.com/`. In that syntax the `TOKEN` part is interpreted as the *username* for HTTP auth, so Git tries to ask for the password, which fails because there's no interactive shell in the workflow. The correct format would be `https://USER:TOKEN@github.com/`, but it's much easier to use [`actions/checkout` with a `path` parameter](https://github.com/actions/checkout#checkout-multiple-repos-private) to clone the repository you want to copy from (and `token`, if necessary). :wink: 

I also strongly recommend avoiding Actions syntax (`${{ ... }}`) in your shell scripts if at all possible. It gets replaced literally to generate the script that's then sent to the shell, so the shell has no idea about the variables. If the values contain any shell control characters, they can break your script in odd ways. For anything already in `env` you can just use shell syntax to access environment variables (e.g. `${TARGET_BRANCH}` instead of `${{ env.TARGET_BRANCH }}`), for secrets you can define an environment variable *for the step* and then do the same. :slightly_smiling_face:

**Reply by lwcorp** [reply]
Thanks, but in the newer version (which has no {{ }} inside the run part) due to the token being commented out the line that runs is `REPO_URL=$SOURCE_REPO` so I don't think it's about TOKEN VS USER:TOKEN.

If you have a version that works (for example, with path, please write it fully and I'll test it.

**Reply by airtower-luna** [reply]
There's the example I linked above, from the `actions/checkout` docs:

```yaml
- name: Checkout private tools
  uses: actions/checkout@v3
  with:
    repository: my-org/my-private-tools
    token: ${{ secrets.GH_PAT }} # `GH_PAT` is a secret that contains your PAT
    path: my-tools
```

If you use `${{ secrets[env.TOKEN_NAME] }}` for the `token` it should just evaluate to an empty string if `TOKEN_NAME` isn't set.

With that (plus adjusted `repository` and `path`, maybe `ref`) you should be able to remove the cloning from your `run` step, and just copy the files you need.

**Reply by lwcorp** [reply]
Could you write the whole code? And what do you put in `on:` so it will run when someone else's repo updates its files?

**Reply by airtower-luna** [reply]
I'm not gonna write a whole workflow, no. The step above would replace almost all of your "Pull File" step, everything above `cd source-repo`. You'd then need another `run` step that copies/moves the files.

> And what do you put in `on:` so it will run when someone else's repo updates its files?

There's no way to do that. Essentially you have three options:

1. If the owner of the other repository cooperates, you could have their repository send a `repository_dispatch` or `workflow_dispatch` to your repository. For that you'd have to give them a token to authorize sending the event.
2. Use a `schedule` event to have your workflow run regularly (e.g. daily or weekly). Check if the files have been updated in the other repository, if yes copy them over.
3. Run the workflow manually (that'd also mean using `workflow_dispatch`).

See [Events that trigger workflows](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows).

**Comment by lwcorp** [comment, marked as answer]
So here's the full code I ended up doing:

 1. Runs manually if needed.
 1. Automatically (can be turned off) monitors your own pushes, only specific branches and folders if needed.
 1. Upon a push grabs another repository's files - it can use tokens for private repos and/or if needed grab specific branches (thanks @airtower-luna!).
 1. It renames the files you want and - only if they're updated - copies them to your own repo.
 1. Only if they're updated - re-commits and re-pushes them to your repo (thankfully [GitHub has loop protection][1]).

The only thing I would hope is to find a way to only define branches and paths for the push monitoring if they're needed instead of commenting them out if they're not needed.

The code:

    name: Pull Files from Another Repository on Push
    env:
      SOURCE_REPO: owner/repo
      AUTOMATIC_MONITOR: true
      FILES: "*.txt,LICENSE" # seperate multiple entries with ,
      FILES_NEW: "*.txt,LICENSE*" # seperate multiple entries with ,
      EXTRA: mv LICENSE LICENSE_some_repo # do an extra command like renaming
      USERNAME: ${{ github.actor }}
      ADDRESS_SUFFIX: users.noreply.github.com
      TOKEN_NAME: ACCESS_TOKEN # name of token defined in the target repo's settings, needed for private repos
      # SOURCE_BRANCH: source_alternate_branch # uncomment to takes files from a non default source branch
      # TARGET_BRANCH: target_alternate_branch # uncomment to monitor a non default target branch
      # TARGET_PATH: target_alternate_path # uncomment to monitor a non default target branch
      THE_SERVER: ${{ github.server_url }}
      PATH_SOURCE_CHECKOUT: temp_folder
      THE_SECRET: ${{ secrets.TOKEN_NAME || 'default_value' }}
    
    on: # Remove # below for non default branches
      workflow_dispatch:
      push:
    #    branches:
    #      - ${{ env.TARGET_BRANCH }}
    #    paths:
    #      - ${{ env.TARGET_PATH }}
    
    jobs:
      pull-file:
        runs-on: ubuntu-latest
        steps:
          - name: Check whether to automatically monitor
            if: ${{ github.event_name != 'workflow_dispatch' && env.AUTOMATIC_MONITOR == false }}
            run: |
              echo "Set not to run automatically. Exiting."
              echo "exiting1=true" >> $GITHUB_ENV

          - name: Checkout
            if: env.exiting1 != 'true'
            uses: actions/checkout@v3
    
          - name: Checkout source with token and branch
            if: env.exiting1 != 'true' && env.THE_SECRET != 'default_value' && env.SOURCE_BRANCH
            uses: actions/checkout@v3
            with:
              repository: ${{ env.SOURCE_REPO }}
              ref: ${{ env.SOURCE_BRANCH }}
              token: ${{ secrets[env.TOKEN_NAME] }}
              path: ${{ env.PATH_SOURCE_CHECKOUT }}
    
          - name: Checkout source with token but without branch
            if: env.exiting1 != 'true' && env.THE_SECRET != 'default_value' && !env.SOURCE_BRANCH
            uses: actions/checkout@v3
            with:
              repository: ${{ env.SOURCE_REPO }}
              token: ${{ secrets[env.TOKEN_NAME] }}
              path: ${{ env.PATH_SOURCE_CHECKOUT }}
    
          - name: Checkout source without token but with branch
            if: env.exiting1 != 'true' && env.THE_SECRET == 'default_value' && env.SOURCE_BRANCH
            uses: actions/checkout@v3
            with:
              repository: ${{ env.SOURCE_REPO }}
              ref: ${{ env.SOURCE_BRANCH }}
              path: ${{ env.PATH_SOURCE_CHECKOUT }}
    
          - name: Checkout source without token and without branch
            if: env.exiting1 != 'true' && env.THE_SECRET == 'default_value' && !env.SOURCE_BRANCH
            uses: actions/checkout@v3
            with:
              repository: ${{ env.SOURCE_REPO }}
              path: ${{ env.PATH_SOURCE_CHECKOUT }}
    
          - name: Update
            if: env.exiting1 != 'true'
            run: |
              cd $PATH_SOURCE_CHECKOUT
              $EXTRA
              cp -u $FILES_NEW ../
              cd ..

          - name: Check for changes
            if: env.exiting1 != 'true'
            run: |
              if git diff --quiet; then
                echo "No changes detected. Exiting."
                echo "exiting2=true" >> $GITHUB_ENV
              fi
              
          - name: Commit
            if: env.exiting1 != 'true' && env.exiting2 != 'true'
            run: |
              git config user.name "$USERNAME"
              git config user.email "$USERNAME@$ADDRESS_SUFFIX"
              git add $FILES_NEW
              git commit -m "Pulled files from $THE_SERVER/$SOURCE_REPO."

          - name: Push
            if: env.exiting1 != 'true' && env.exiting2 != 'true'
            run: |
              if [ -n "$TARGET_BRANCH" ]; then
                git push origin $TARGET_BRANCH
              else
                git push
              fi


  [1]: https://github.com/orgs/community/discussions/26970#discussioncomment-3254152
